### ISSUE TITLE
Pinning a specific package

### ISSUE THREAD

**Issue by domenkozar** [cachix/devenv#16]
This is a common use case. Please describe below if we're missing a use case.

- picking up language specific files (`Gemspec`, ..): we do that for some languages as Ruby and Rust, more support need to be implemented.
- using older version because there's an issue with the one shipped in nixpkgs

Docs: https://devenv.sh/recipes/nix/#getting-a-recent-version-of-a-package-from-nixpkgs-unstable

**Comment by roman** [comment]
Have you considered reading language-specific files that indicate the desired language version? (e.g. `go.mod`, `Gemspec`, etc.)?

**Comment by domenkozar** [comment]
> Have you considered reading language-specific files that indicate the desired language version? (e.g. `go.mod`, `Gemspec`, etc.)?

Yes, see https://github.com/cachix/devenv/pull/401 how it's implemented for Ruby

**Comment by worldofgeese** [comment]
@domenkozar could adopting something like flox's [channels](https://floxdev.com/docs/reference/command-reference/flox-channels/) or Devbox's newly announced [Nixhub](https://www.jetpack.io/blog/introducing-nixhub/) work to provide historical versions to `devenv`? I ran into this recently with `bun`, where I needed an older version than 0.7.0 provided in unstable. I would have used the method provided in https://github.com/cachix/devenv/issues/218#issuecomment-1365842092 but https://lazamar.co.uk/nix-versions/?channel=nixpkgs-unstable&package=bun doesn't show more recent versions I know exist in [merged PRs to Nixpkgs](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+bun&type=pullrequests).

**Comment by domenkozar** [comment]
I made a prototype for this back in the beginning of the year, still plan to finish it.

**Comment by vic** [comment]
Hello, 

Just recently I made a tool for myself [nix-versions](https://nix-versions.alwaysdata.net) ([source](https://github.com/vic/nix-versions)) that works as a front-end CLI for either Lazamar and NixHub version indexes.

Today I wrote a bit more of documentation and deployed a server for it. The server also provides a [flake generator endpoint](https://nix-versions.alwaysdata.net/flake-generator.html), allowing people to generate flakes for version-pinned packages. I believe that is a common feature people want (#1732) not only from devenv, but from any flake (NixOS configurations, etc) where people need better guarantees that some particular packages (like compilers/interpreters) are at known compatible releases that work with their code (I'm looking at you Scala!). At the end of that page, I mention how to add nix-versions generated flakes as [devenv inputs](https://nix-versions.alwaysdata.net/flake-generator.html#using-as-input-on-devenv-environments)

**Comment by itpropro** [comment]
This is very essential for every developer environment. It is not really practical to pin a separate version of nixpkgs for every different version of a tool needed in a dev environment. Normally we have at least 20-30 different tools all with a specific pinned version that we would want to specify.

**Comment by domenkozar** [comment]
> This is very essential for every developer environment. It is not really practical to pin a separate version of nixpkgs for every different version of a tool needed in a dev environment. Normally we have at least 20-30 different tools all with a specific pinned version that we would want to specify.

How comes you want to pin each tool? Each tool used (latest) is already pinned using `devenv.lock`.

**Comment by itpropro** [comment]
Because every project uses other versions of the same tools. I have never seen 2 projects that have the same version of any tool. This can have a lot of reasons from compliance and security to compatibility, but in reality, tools almost never use latest.
There are a dozen of examples for dev tools, but it should already be enough to just look at the different versions of python and python packages that are in nixpkgs. The whole python package system is so brittle that you rarely touch a working combination of package versions. 
In addition to that with the rise of supply chain attacks, every new package version needs to be verified and approved in most projects I work in. So a normal flow would be to approve one new version of one tool, bring it into the stack and keep all others the same version as before. Then maybe decide to skip the version bump for tool 2,3 and 4 and for tool 5 we do the same process again. Sometimes there a weeks between these processes, which would mean we have to pin a specific nixpkgs version for every single tool to have that exact version pinned.

**Comment by fzakaria** [comment]
Checkout: https://fzakaria.com/2026/08/09/nixpkgs-multiverse-every-version-that-ever-existed
That should solve the need :)

**Comment by domenkozar** [comment]
THIS IS FINALLY FIXED!

https://devenv.sh/pinning/#pinning-an-individual-package-version

**Comment by assyrus-favolo** [comment]
Amazing, thanks a ton!
<img width="200" height="134" alt="Image" src="https://github.com/user-attachments/assets/139c053f-458b-4615-9a54-f6c0d6853e86" />
